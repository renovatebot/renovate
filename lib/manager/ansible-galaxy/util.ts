export const newBlockRegEx = /^\s*-\s*((\w+):\s*(.*))$/;
export const blockLineRegEx = /^\s*((\w+):\s*(.*))$/;
export const galaxyDepRegex = /[\w-]+\.[\w-]+/;
export const dependencyRegex = /^dependencies:/;
export const galaxyRegEx =
  /^\s+(?<lookupName>[\w.]+):\s*["'](?<version>.+)["']\s*/;
export const nameMatchRegex =
  /(?<source>((git\+)?(?:(git|ssh|https?):\/\/)?(.*@)?(?<hostname>[\w.-]+)(?:(:\d+)?\/|:))(?<depName>[\w./-]+)(?:\.git)?)(,(?<version>[\w.]*))?/;
