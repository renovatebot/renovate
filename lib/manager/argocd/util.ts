export const keyValueExtractionRegex =
  /^\s*(?<key>[^\s]+):\s+"?(?<value>[^"\s]+)"?\s*$/;
// looks for `apiVersion: argoproj.io/
export const fileTestRegex = /\s*apiVersion:\s*argoproj.io\/\s*/;
