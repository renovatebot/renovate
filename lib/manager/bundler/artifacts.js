/* istanbul ignore file */

const { exec } = require('child-process-promise');
const fs = require('fs-extra');
const upath = require('upath');

const { getPkgReleases } = require('../../datasource/docker');
const {
  isValid,
  isVersion,
  matches,
  sortVersions,
} = require('../../versioning/ruby');

module.exports = {
  getArtifacts,
};

async function getArtifacts(
  packageFileName,
  updatedDeps,
  newPackageFileContent,
  config
) {
  logger.debug(`bundler.getArtifacts(${packageFileName})`);
  // istanbul ignore if
  if (global.repoCache.bundlerArtifactsError) {
    logger.info('Aborting Bundler artifacts due to previous failed attempt');
    throw new Error(global.repoCache.bundlerArtifactsError);
  }
  const lockFileName = packageFileName + '.lock';
  const existingLockFileContent = await platform.getFile(lockFileName);
  if (!existingLockFileContent) {
    logger.debug('No Gemfile.lock found');
    return null;
  }
  const cwd = upath.join(config.localDir, upath.dirname(packageFileName));
  let stdout;
  let stderr;
  try {
    const localPackageFileName = upath.join(config.localDir, packageFileName);
    await fs.outputFile(localPackageFileName, newPackageFileContent);
    const localLockFileName = upath.join(config.localDir, lockFileName);
    if (!config.gitFs) {
      await fs.outputFile(localLockFileName, existingLockFileContent);
      const fileList = await platform.getFileList();
      const gemspecs = fileList.filter(file => file.endsWith('.gemspec'));
      for (const gemspec of gemspecs) {
        const content = await platform.getFile(gemspec);
        await fs.outputFile(upath.join(config.localDir, gemspec), content);
      }
    }
    const env =
      global.trustLevel === 'high'
        ? process.env
        : {
            HOME: process.env.HOME,
            PATH: process.env.PATH,
          };
    const startTime = process.hrtime();
    let cmd;
    if (config.binarySource === 'docker') {
      logger.info('Running bundler via docker');
      let tag = 'latest';
      let rubyConstraint;
      const rubyVersionFile = upath.join(
        upath.dirname(packageFileName),
        '.ruby-version'
      );
      logger.debug('Checking ' + rubyVersionFile);
      const rubyVersionFileContent = await platform.getFile(rubyVersionFile);
      if (rubyVersionFileContent) {
        logger.debug('Using ruby version specified in .ruby-version');
        rubyConstraint = rubyVersionFileContent.replace(/\n/g, '').trim();
      } else {
        rubyConstraint =
          config && config.compatibility && config.compatibility.ruby
            ? config.compatibility.ruby
            : undefined;
      }
      if (rubyConstraint && isValid(rubyConstraint)) {
        logger.debug('Found ruby compatibility');
        const rubyReleases = await getPkgReleases({
          lookupName: 'renovate/ruby',
        });
        if (rubyReleases && rubyReleases.releases) {
          let versions = rubyReleases.releases.map(release => release.version);
          versions = versions.filter(version => isVersion(version));
          versions = versions.filter(version =>
            matches(version, rubyConstraint)
          );
          versions = versions.sort(sortVersions);
          if (versions.length) {
            tag = versions.pop();
          }
        }
      }
      const bundlerConstraint =
        config && config.compatibility && config.compatibility.bundler
          ? config.compatibility.bundler
          : undefined;
      let bundlerVersion = '';
      if (bundlerConstraint && isVersion(bundlerConstraint)) {
        bundlerVersion = ' -v ' + bundlerConstraint;
      }
      cmd = `docker run --rm `;
      const volumes = [config.localDir];
      cmd += volumes.map(v => `-v ${v}:${v} `).join('');
      const envVars = [];
      cmd += envVars.map(e => `-e ${e} `);
      cmd += `-w ${cwd} `;
      cmd += `renovate/ruby:${tag} bash -l -c "ruby --version && `;
      cmd += 'gem install bundler' + bundlerVersion;
      cmd += ' && bundle';
    } else {
      logger.info('Running bundler via global bundler');
      cmd = 'bundler';
    }
    cmd += ' lock"';
    logger.debug({ cmd }, 'bundler command');
    ({ stdout, stderr } = await exec(cmd, {
      cwd,
      shell: true,
      env,
    }));
    const duration = process.hrtime(startTime);
    const seconds = Math.round(duration[0] + duration[1] / 1e9);
    logger.info(
      { seconds, type: 'Gemfile.lock', stdout, stderr },
      'Generated lockfile'
    );
    // istanbul ignore if
    if (config.gitFs) {
      const status = await platform.getRepoStatus();
      if (!status.modified.includes(lockFileName)) {
        return null;
      }
    } else {
      const newLockFileContent = await fs.readFile(localLockFileName, 'utf8');

      if (newLockFileContent === existingLockFileContent) {
        logger.debug('Gemfile.lock is unchanged');
        return null;
      }
    }
    logger.debug('Returning updated Gemfile.lock');
    return [
      {
        file: {
          name: lockFileName,
          contents: await fs.readFile(localLockFileName, 'utf8'),
        },
      },
    ];
  } catch (err) {
    if (
      err.stdout &&
      err.stdout.includes('No such file or directory') &&
      !config.gitFs
    ) {
      logger.warn(
        { err },
        'It is necessary to run Renovate in gitFs mode - contact your bot administrator'
      );
      global.repoCache.bundlerArtifactsError = 'bundler-fs';
      throw new Error('bundler-fs');
    }
    if (
      (err.stdout &&
        err.stdout.includes('Please supply credentials for this source')) ||
      (err.stderr && err.stderr.includes('Authentication is required'))
    ) {
      logger.warn(
        { err },
        'Gemfile.lock update failed due to missing credentials'
      );
      global.repoCache.bundlerArtifactsError = 'bundler-credentials';
      throw new Error('bundler-credentials');
    }
    logger.warn(
      { err, message: err.message },
      'Failed to generate Gemfile.lock (unknown error)'
    );
    global.repoCache.bundlerArtifactsError = 'bundler-unknown';
    throw new Error('bundler-unknown');
  }
}
