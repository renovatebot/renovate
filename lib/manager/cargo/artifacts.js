const upath = require('upath');
const process = require('process');
const fs = require('fs-extra');
const { exec } = require('child-process-promise');

module.exports = {
  updateArtifacts,
};

async function updateArtifacts(
  packageFileName,
  updatedDeps,
  newPackageFileContent,
  config
) {
  await logger.debug(`cargo.updateArtifacts(${packageFileName})`);
  if (updatedDeps === undefined || updatedDeps.length < 1) {
    logger.debug('No updated cargo deps - returning null');
    return null;
  }
  const lockFileName = 'Cargo.lock';
  const existingLockFileContent = await platform.getFile(lockFileName);
  if (!existingLockFileContent) {
    logger.debug('No Cargo.lock found');
    return null;
  }
  const localPackageFileName = upath.join(config.localDir, packageFileName);
  const localLockFileName = upath.join(config.localDir, lockFileName);
  let stdout;
  let stderr;
  try {
    await fs.outputFile(localPackageFileName, newPackageFileContent);
    logger.debug('Updating ' + lockFileName);
    const cwd = config.localDir;
    const env =
      global.trustLevel === 'high'
        ? process.env
        : {
            HOME: process.env.HOME,
            PATH: process.env.PATH,
          };
    for (let i = 0; i < updatedDeps.length; i += 1) {
      const dep = updatedDeps[i];
      // Update dependency `${dep}` in Cargo.lock file corresponding to Cargo.toml file located
      // at ${localPackageFileName} path
      let cmd;
      // istanbul ignore if
      if (config.binarySource === 'docker') {
        logger.info('Running cargo via docker');
        cmd = `docker run --rm `;
        const volumes = [cwd];
        cmd += volumes.map(v => `-v ${v}:${v} `).join('');
        const envVars = [];
        cmd += envVars.map(e => `-e ${e} `);
        cmd += `-w ${cwd} `;
        cmd += `renovate/rust cargo`;
      } else {
        logger.info('Running cargo via global cargo');
        cmd = 'cargo';
      }
      cmd += ` update --manifest-path ${localPackageFileName} --package ${dep}`;
      const startTime = process.hrtime();
      try {
        ({ stdout, stderr } = await exec(cmd, {
          cwd,
          shell: true,
          env,
        }));
      } catch (err) /* istanbul ignore next */ {
        // Two different versions of one dependency can be present in the same
        // crate, and when that happens an attempt to update it with --package ${dep}
        // key results in cargo exiting with error code `101` and an error mssage:
        // "error: There are multiple `${dep}` packages in your project".
        //
        // If exception `err` was caused by this, we execute `updateAll` function
        // instead of returning an error. `updateAll` function just executes
        // "cargo update --manifest-path ${localPackageFileName}" without the `--package` key.
        //
        // If exception `err` was not caused by this, we just rethrow it. It will be caught
        // by the outer try { } catch {} and processed normally.
        const msgStart = 'error: There are multiple';
        if (err.code === 101 && err.stderr.startsWith(msgStart)) {
          cmd = cmd.replace(/ --package.*/, '');
          ({ stdout, stderr } = await exec(cmd, {
            cwd,
            shell: true,
            env,
          }));
        } else {
          throw err; // this is caught below
        }
      }
      const duration = process.hrtime(startTime);
      const seconds = Math.round(duration[0] + duration[1] / 1e9);
      logger.info(
        { seconds, type: 'Cargo.lock', stdout, stderr },
        'Updated lockfile'
      );
    }
    logger.debug('Returning updated Cargo.lock');
    const newCargoLockContent = await fs.readFile(localLockFileName, 'utf8');
    if (existingLockFileContent === newCargoLockContent) {
      logger.debug('Cargo.lock is unchanged');
      return null;
    }
    return [
      {
        file: {
          name: lockFileName,
          contents: newCargoLockContent,
        },
      },
    ];
  } catch (err) {
    logger.warn({ err }, 'Failed to update Cargo lock file');
    return {
      lockFileError: {
        lockFile: lockFileName,
        stderr: err.message,
      },
    };
  }
}
