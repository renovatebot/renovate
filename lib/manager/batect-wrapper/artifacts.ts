import { logger } from '../../logger';
import { Http } from '../../util/http';
import type { UpdateArtifact, UpdateArtifactsResult } from '../types';

const http = new Http('batect-wrapper');

async function updateArtifact(
  path: string,
  fileName: string,
  version: string
): Promise<UpdateArtifactsResult> {
  const url = `https://github.com/batect/batect/releases/download/${version}/${fileName}`;

  try {
    const response = await http.get(url);
    const contents = response.body;

    return {
      file: {
        name: path,
        contents,
      },
    };
  } catch (err) {
    const errorDescription: string = err.toString();

    return {
      artifactError: {
        lockFile: path,
        stderr: `HTTP GET ${url} failed: ${errorDescription}`,
      },
    };
  }
}

export async function updateArtifacts({
  packageFileName,
  config,
}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
  const version = config.newVersion;

  logger.debug({ version, packageFileName }, 'Updating Batect wrapper scripts');

  return [
    await updateArtifact(packageFileName, 'batect', version),
    await updateArtifact(`${packageFileName}.cmd`, 'batect.cmd', version),
  ];
}
