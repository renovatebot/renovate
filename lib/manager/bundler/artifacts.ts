import {
  getSiblingFileName,
  readLocalFile,
  writeLocalFile,
} from '../../util/fs';
import { exec, ExecOptions } from '../../util/exec';
import { logger } from '../../logger';
import { isValid } from '../../versioning/ruby';
import { UpdateArtifact, UpdateArtifactsResult } from '../common';
import { platform } from '../../platform';
import {
  BUNDLER_INVALID_CREDENTIALS,
  BUNDLER_UNKNOWN_ERROR,
} from '../../constants/error-messages';
import { HostRule } from '../../types';
import {
  getAuthenticationHeaderValue,
  findAllAuthenticatable,
  getDomain,
} from './host-rules';

const hostConfigVariablePrefix = 'BUNDLE_';

async function getRubyConstraint(
  updateArtifact: UpdateArtifact
): Promise<string> {
  const { packageFileName, config } = updateArtifact;
  const { compatibility = {} } = config;
  const { ruby } = compatibility;

  let rubyConstraint: string;
  if (ruby) {
    logger.debug('Using rubyConstraint from config');
    rubyConstraint = ruby;
  } else {
    const rubyVersionFile = getSiblingFileName(
      packageFileName,
      '.ruby-version'
    );
    const rubyVersionFileContent = await platform.getFile(rubyVersionFile);
    if (rubyVersionFileContent) {
      logger.debug('Using ruby version specified in .ruby-version');
      rubyConstraint = rubyVersionFileContent
        .replace(/^ruby-/, '')
        .replace(/\n/g, '')
        .trim();
    }
  }
  return rubyConstraint;
}

function buildBundleHostVariable(hostRule: HostRule): Record<string, string> {
  const varName =
    hostConfigVariablePrefix +
    getDomain(hostRule)
      .split('.')
      .map(term => term.toUpperCase())
      .join('__');

  return {
    [varName]: `${getAuthenticationHeaderValue(hostRule)}`,
  };
}

export async function updateArtifacts(
  updateArtifact: UpdateArtifact
): Promise<UpdateArtifactsResult[] | null> {
  const {
    packageFileName,
    updatedDeps,
    newPackageFileContent,
    config,
  } = updateArtifact;
  const { compatibility = {} } = config;

  logger.debug(`bundler.updateArtifacts(${packageFileName})`);
  // istanbul ignore if
  if (global.repoCache.bundlerArtifactsError) {
    logger.debug('Aborting Bundler artifacts due to previous failed attempt');
    throw new Error(global.repoCache.bundlerArtifactsError);
  }
  const lockFileName = `${packageFileName}.lock`;
  const existingLockFileContent = await platform.getFile(lockFileName);
  if (!existingLockFileContent) {
    logger.debug('No Gemfile.lock found');
    return null;
  }
  try {
    await writeLocalFile(packageFileName, newPackageFileContent);

    const cmd = `bundle lock --update ${updatedDeps.join(' ')}`;

    let bundlerVersion = '';
    const { bundler } = compatibility;
    if (bundler) {
      if (isValid(bundler)) {
        logger.debug({ bundlerVersion: bundler }, 'Found bundler version');
        bundlerVersion = ` -v ${bundler}`;
      } else {
        logger.warn({ bundlerVersion: bundler }, 'Invalid bundler version');
      }
    } else {
      logger.debug('No bundler version constraint found - will use latest');
    }
    const preCommands = [
      'ruby --version',
      `gem install bundler${bundlerVersion} --no-document`,
    ];

    const bundlerHostRulesVariables = findAllAuthenticatable({
      hostType: 'bundler',
    }).reduce((variables, hostRule) => {
      return { ...variables, ...buildBundleHostVariable(hostRule) };
    }, {} as Record<string, string>);

    const execOptions: ExecOptions = {
      cwdFile: packageFileName,
      extraEnv: bundlerHostRulesVariables,
      docker: {
        image: 'renovate/ruby',
        tagScheme: 'ruby',
        tagConstraint: await getRubyConstraint(updateArtifact),
        preCommands,
      },
    };
    await exec(cmd, execOptions);
    const status = await platform.getRepoStatus();
    if (!status.modified.includes(lockFileName)) {
      return null;
    }
    logger.debug('Returning updated Gemfile.lock');
    const lockFileContent = await readLocalFile(lockFileName);
    return [
      {
        file: {
          name: lockFileName,
          contents: lockFileContent,
        },
      },
    ];
  } catch (err) /* istanbul ignore next */ {
    const output = err.stdout + err.stderr;
    if (
      err.message.includes('fatal: Could not parse object') ||
      output.includes('but that version could not be found')
    ) {
      return [
        {
          artifactError: {
            lockFile: lockFileName,
            stderr: output,
          },
        },
      ];
    }
    if (
      (err.stdout &&
        err.stdout.includes('Please supply credentials for this source')) ||
      (err.stderr && err.stderr.includes('Authentication is required')) ||
      (err.stderr &&
        err.stderr.includes(
          'Please make sure you have the correct access rights'
        ))
    ) {
      logger.debug(
        { err },
        'Gemfile.lock update failed due to missing credentials - skipping branch'
      );
      // Do not generate these PRs because we don't yet support Bundler authentication
      global.repoCache.bundlerArtifactsError = BUNDLER_INVALID_CREDENTIALS;
      throw new Error(BUNDLER_INVALID_CREDENTIALS);
    }
    const resolveMatchRe = new RegExp('\\s+(.*) was resolved to', 'g');
    if (output.match(resolveMatchRe)) {
      logger.debug({ err }, 'Bundler has a resolve error');
      const resolveMatches = [];
      let resolveMatch;
      do {
        resolveMatch = resolveMatchRe.exec(output);
        if (resolveMatch) {
          resolveMatches.push(resolveMatch[1].split(' ').shift());
        }
      } while (resolveMatch);
      if (resolveMatches.some(match => !updatedDeps.includes(match))) {
        logger.debug(
          { resolveMatches, updatedDeps },
          'Found new resolve matches - reattempting recursively'
        );
        const newUpdatedDeps = [
          ...new Set([...updatedDeps, ...resolveMatches]),
        ];
        return updateArtifacts({
          packageFileName,
          updatedDeps: newUpdatedDeps,
          newPackageFileContent,
          config,
        });
      }
      logger.debug(
        { err },
        'Gemfile.lock update failed due to incompatible packages'
      );
      return [
        {
          artifactError: {
            lockFile: lockFileName,
            stderr: err.stdout + '\n' + err.stderr,
          },
        },
      ];
    }
    logger.warn(
      { err },
      'Gemfile.lock update failed due to unknown reason - skipping branch'
    );
    global.repoCache.bundlerArtifactsError = BUNDLER_UNKNOWN_ERROR;
    throw new Error(BUNDLER_UNKNOWN_ERROR);
  }
}
