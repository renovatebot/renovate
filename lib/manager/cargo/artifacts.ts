import { quote } from 'shlex';
import { logger } from '../../logger';
import { ExecOptions, exec } from '../../util/exec';
import {
  findLocalSiblingOrParent,
  readLocalFile,
  writeLocalFile,
} from '../../util/fs';
import { UpdateArtifact, UpdateArtifactsResult } from '../common';

async function cargoUpdate(
  manifestPath: string,
  packageName?: string
): Promise<void> {
  let cmd = `cargo update --manifest-path ${quote(manifestPath)}`;
  if (packageName) {
    cmd += ` --package ${quote(packageName)}`;
  }

  const execOptions: ExecOptions = {
    docker: {
      image: 'renovate/rust',
    },
  };
  try {
    await exec(cmd, execOptions);
  } catch (err) /* istanbul ignore next */ {
    // Two different versions of one dependency can be present in the same
    // crate, and when that happens an attempt to update it with --package ${dep}
    // key results in cargo exiting with error code `101` and an error message:
    // "error: There are multiple `${dep}` packages in your project".
    //
    // If exception `err` was caused by this, we execute `updateAll` function
    // instead of returning an error. `updateAll` function just executes
    // "cargo update --manifest-path ${localPackageFileName}" without the `--package` key.
    //
    // If exception `err` was not caused by this, we just rethrow it. It will be caught
    // by the outer try { } catch {} and processed normally.
    const msgStart = 'error: There are multiple';
    if (err.code === 101 && err.stderr.startsWith(msgStart)) {
      cmd = cmd.replace(/ --package.*/, '');
      await exec(cmd, execOptions);
    } else {
      throw err; // this is caught below
    }
  }
}

export async function updateArtifacts({
  packageFileName,
  updatedDeps,
  newPackageFileContent,
  config,
}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
  logger.debug(`cargo.updateArtifacts(${packageFileName})`);

  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';

  if (
    !isLockFileMaintenance &&
    (updatedDeps === undefined || updatedDeps.length < 1)
  ) {
    logger.debug('No updated cargo deps - returning null');
    return null;
  }

  // For standalone package crates, the `Cargo.lock` will be in the same
  // directory as `Cargo.toml` (ie. a sibling). For cargo workspaces, it
  // will be further up.
  const lockFileName = await findLocalSiblingOrParent(
    packageFileName,
    'Cargo.lock'
  );
  const existingLockFileContent = lockFileName
    ? await readLocalFile(lockFileName)
    : null;
  if (!existingLockFileContent) {
    logger.debug('No Cargo.lock found');
    return null;
  }
  try {
    await writeLocalFile(packageFileName, newPackageFileContent);
    logger.debug('Updating ' + lockFileName);
    for (let i = 0; i < updatedDeps.length; i += 1) {
      const dep = updatedDeps[i];
      // Update dependency `${dep}` in Cargo.lock file corresponding to Cargo.toml file located
      // at ${localPackageFileName} path
      await cargoUpdate(packageFileName, dep);
    }
    if (isLockFileMaintenance) {
      await cargoUpdate(packageFileName);
    }
    logger.debug('Returning updated Cargo.lock');
    const newCargoLockContent = await readLocalFile(lockFileName);
    if (existingLockFileContent === newCargoLockContent) {
      logger.debug('Cargo.lock is unchanged');
      return null;
    }
    return [
      {
        file: {
          name: lockFileName,
          contents: newCargoLockContent,
        },
      },
    ];
  } catch (err) {
    logger.warn({ err }, 'Failed to update Cargo lock file');
    return [
      {
        artifactError: {
          lockFile: lockFileName,
          stderr: err.message,
        },
      },
    ];
  }
}
