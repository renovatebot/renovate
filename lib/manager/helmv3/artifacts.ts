import { quote } from 'shlex';
import { TEMPORARY_ERROR } from '../../constants/error-messages';
import { logger } from '../../logger';
import { ExecOptions, exec } from '../../util/exec';
import {
  getSiblingFileName,
  getSubDirectory,
  readLocalFile,
  writeLocalFile,
} from '../../util/fs';
import type { UpdateArtifact, UpdateArtifactsResult } from '../types';

async function helmUpdate(manifestPath: string): Promise<void> {
  const cmd = `helm dependency update ${quote(getSubDirectory(manifestPath))}`;

  const execOptions: ExecOptions = {
    docker: {
      image: 'helm',
    },
  };
  await exec(cmd, execOptions);
}

export async function updateArtifacts({
  packageFileName,
  updatedDeps,
  newPackageFileContent,
  config,
}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
  logger.debug(`helmv3.updateArtifacts(${packageFileName})`);

  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';

  if (
    !isLockFileMaintenance &&
    (updatedDeps === undefined || updatedDeps.length < 1)
  ) {
    logger.debug('No updated helmv3 deps - returning null');
    return null;
  }

  const lockFileName = getSiblingFileName(packageFileName, 'Chart.lock');
  const existingLockFileContent = await readLocalFile(lockFileName);
  if (!existingLockFileContent) {
    logger.debug('No Chart.lock found');
    return null;
  }
  try {
    await writeLocalFile(packageFileName, newPackageFileContent);
    logger.debug('Updating ' + lockFileName);
    await helmUpdate(packageFileName);
    logger.debug('Returning updated Chart.lock');
    const newHelmLockContent = await readLocalFile(lockFileName);
    if (existingLockFileContent === newHelmLockContent) {
      logger.debug('Chart.lock is unchanged');
      return null;
    }
    return [
      {
        file: {
          name: lockFileName,
          contents: newHelmLockContent,
        },
      },
    ];
  } catch (err) {
    // istanbul ignore if
    if (err.message === TEMPORARY_ERROR) {
      throw err;
    }
    logger.warn({ err }, 'Failed to update Helm lock file');
    return [
      {
        artifactError: {
          lockFile: lockFileName,
          stderr: err.message,
        },
      },
    ];
  }
}
