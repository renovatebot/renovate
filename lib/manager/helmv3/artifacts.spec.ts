import { exec as _exec } from 'child_process';
import _fs from 'fs-extra';
import { join } from 'upath';
import { envMock, mockExecAll } from '../../../test/execUtil';
import { git, mocked } from '../../../test/util';
import { setExecConfig } from '../../util/exec';
import { BinarySource } from '../../util/exec/common';
import * as docker from '../../util/exec/docker';
import * as _env from '../../util/exec/env';
import * as helmv3 from './artifacts';

jest.mock('fs-extra');
jest.mock('child_process');
jest.mock('../../util/exec/env');
jest.mock('../../util/git');
jest.mock('../../util/http');

const fs: jest.Mocked<typeof _fs> = _fs as any;
const exec: jest.Mock<typeof _exec> = _exec as any;
const env = mocked(_env);

const config = {
  // `join` fixes Windows CI
  localDir: join('/tmp/github/some/repo'),
  dockerUser: 'foobar',
};

describe('.updateArtifacts()', () => {
  beforeEach(async () => {
    jest.resetAllMocks();
    jest.resetModules();

    env.getChildProcessEnv.mockReturnValue(envMock.basic);
    await setExecConfig(config);
    docker.resetPrefetchedImages();
  });
  it('returns null if no Chart.lock found', async () => {
    const updatedDeps = ['dep1'];
    expect(
      await helmv3.updateArtifacts({
        packageFileName: 'Chart.yaml',
        updatedDeps,
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
  });
  it('returns null if updatedDeps is empty', async () => {
    expect(
      await helmv3.updateArtifacts({
        packageFileName: 'Chart.yaml',
        updatedDeps: [],
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
  });
  it('returns null if unchanged', async () => {
    fs.readFile.mockResolvedValueOnce('Current Chart.lock' as any);
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('Current Chart.lock' as any);
    const updatedDeps = ['dep1'];
    expect(
      await helmv3.updateArtifacts({
        packageFileName: 'Chart.yaml',
        updatedDeps,
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });
  it('returns updated Chart.lock', async () => {
    git.getFile.mockResolvedValueOnce('Old Chart.lock');
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('New Chart.lock' as any);
    const updatedDeps = ['dep1'];
    expect(
      await helmv3.updateArtifacts({
        packageFileName: 'Chart.yaml',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('returns updated Chart.lock for lockfile maintenance', async () => {
    git.getFile.mockResolvedValueOnce('Old Chart.lock');
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('New Chart.lock' as any);
    expect(
      await helmv3.updateArtifacts({
        packageFileName: 'Chart.yaml',
        updatedDeps: [],
        newPackageFileContent: '{}',
        config: { ...config, updateType: 'lockFileMaintenance' },
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('returns updated Chart.lock with docker', async () => {
    jest.spyOn(docker, 'removeDanglingContainers').mockResolvedValueOnce();
    await setExecConfig({ ...config, binarySource: BinarySource.Docker });
    git.getFile.mockResolvedValueOnce('Old Chart.lock');
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('New Chart.lock' as any);
    const updatedDeps = ['dep1'];
    expect(
      await helmv3.updateArtifacts({
        packageFileName: 'Chart.yaml',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });
  it('catches errors', async () => {
    fs.readFile.mockResolvedValueOnce('Current Chart.lock' as any);
    fs.outputFile.mockImplementationOnce(() => {
      throw new Error('not found');
    });
    const updatedDeps = ['dep1'];
    expect(
      await helmv3.updateArtifacts({
        packageFileName: 'Chart.yaml',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).toMatchSnapshot();
  });
});
