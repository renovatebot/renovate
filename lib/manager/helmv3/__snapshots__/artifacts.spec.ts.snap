// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.updateArtifacts() catches errors 1`] = `
Array [
  Object {
    "artifactError": Object {
      "lockFile": "Chart.lock",
      "stderr": "not found",
    },
  },
]
`;

exports[`.updateArtifacts() returns null if unchanged 1`] = `
Array [
  Object {
    "cmd": "helm dependency update ''",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`.updateArtifacts() returns updated Chart.lock 1`] = `
Array [
  Object {
    "cmd": "helm dependency update ''",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`.updateArtifacts() returns updated Chart.lock for lockfile maintenance 1`] = `
Array [
  Object {
    "cmd": "helm dependency update ''",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`.updateArtifacts() returns updated Chart.lock with docker 1`] = `
Array [
  Object {
    "cmd": "docker pull renovate/helm",
    "options": Object {
      "encoding": "utf-8",
    },
  },
  Object {
    "cmd": "docker ps --filter name=renovate_helm -aq",
    "options": Object {
      "encoding": "utf-8",
    },
  },
  Object {
    "cmd": "docker run --rm --name=renovate_helm --label=renovate_child --user=foobar -v \\"/tmp/github/some/repo\\":\\"/tmp/github/some/repo\\" -w \\"/tmp/github/some/repo\\" renovate/helm bash -l -c \\"helm dependency update ''\\"",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;
